# AI-based Medical Detection and Diagnosis System 

## Describtion
Our system is a powerful platform designed to assist users in understanding and managing their health. Leveraging the capabilities of machine learning, our system analyzes user-input symptoms to predict potential diseases accurately.

### Importing Required Libraires

In [5]:
import  pandas as pd
import numpy as np
import pickle
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix


### Dataset Loading and Displaying

In [6]:
#med_dataset stands for medical dataset
med_dataset = pd.read_csv('Dataset/Training.csv')
med_dataset.shape[1] #number of features in dataset

133

the dataset has 133 input features and each feature represents a symptom.<br>
the feature can be 0 or 1. <br>
if 0 that means the user does not suffer from this symptom.<br>
if 1 that means the user suffers from this symptom.<br>

In [7]:
#Displaying the dataset
med_dataset.head()

,itching,skin_rash,nodal_skin_eruptions,continuous_sneezing,shivering,chills,joint_pain,stomach_pain,acidity,ulcers_on_tongue,...,blackheads,scurring,skin_peeling,silver_like_dusting,small_dents_in_nails,inflammatory_nails,blister,red_sore_around_nose,yellow_crust_ooze,prognosis
0,1,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Fungal infection
1,0,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Fungal infection
2,1,0,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Fungal infection
3,1,1,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Fungal infection
4,1,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,Fungal infection


this is a labled dataset, the last column contains the symptom name.<br>
the models we will use for this dataset are SVM , Random forrest.

### Data Preparation (Exploration and Cleansing)

In [8]:
#displaying the basic information about dataset
med_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4920 entries, 0 to 4919
Columns: 133 entries, itching to prognosis
dtypes: int64(132), object(1)
memory usage: 5.0+ MB


In [9]:
#count the number of values for each features
med_dataset.count()

itching                 4920
skin_rash               4920
nodal_skin_eruptions    4920
continuous_sneezing     4920
shivering               4920
                        ... 
inflammatory_nails      4920
blister                 4920
red_sore_around_nose    4920
yellow_crust_ooze       4920
prognosis               4920
Length: 133, dtype: int64

In [10]:
#showing the datatype of values for each feature
med_dataset.dtypes

itching                  int64
skin_rash                int64
nodal_skin_eruptions     int64
continuous_sneezing      int64
shivering                int64
                         ...  
inflammatory_nails       int64
blister                  int64
red_sore_around_nose     int64
yellow_crust_ooze        int64
prognosis               object
Length: 133, dtype: object

In [11]:
#count how many null values for each feature
med_dataset.isnull().sum()

itching                 0
skin_rash               0
nodal_skin_eruptions    0
continuous_sneezing     0
shivering               0
                       ..
inflammatory_nails      0
blister                 0
red_sore_around_nose    0
yellow_crust_ooze       0
prognosis               0
Length: 133, dtype: int64

this dataset was manully cleaned so, it looks very clean and ready for training.

### Dataset Splitting (Traing Set , Test Set)

In [12]:
#splitting the dataset into input feature and the output
X = med_dataset.drop('prognosis', axis=1) # Removes the target column from the feature set
Y = med_dataset['prognosis'] # Stores the target labels separately

In [13]:
X

,itching,skin_rash,nodal_skin_eruptions,continuous_sneezing,shivering,chills,joint_pain,stomach_pain,acidity,ulcers_on_tongue,...,pus_filled_pimples,blackheads,scurring,skin_peeling,silver_like_dusting,small_dents_in_nails,inflammatory_nails,blister,red_sore_around_nose,yellow_crust_ooze
0,1,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,1,0,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,1,1,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,1,1,1,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4915,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4916,0,1,0,0,0,0,0,0,0,0,...,1,1,1,0,0,0,0,0,0,0
4917,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4918,0,1,0,0,0,0,1,0,0,0,...,0,0,0,1,1,1,1,0,0,0


In [14]:
Y

0                              Fungal infection
1                              Fungal infection
2                              Fungal infection
3                              Fungal infection
4                              Fungal infection
                         ...                   
4915    (vertigo) Paroymsal  Positional Vertigo
4916                                       Acne
4917                    Urinary tract infection
4918                                  Psoriasis
4919                                   Impetigo
Name: prognosis, Length: 4920, dtype: object

In [15]:
#encoding the target variable prognosis using scikit-learn’s LabelEncoder.

# Initialize the encoder
le = LabelEncoder()

# Fit on the target values
le.fit(Y)

# Transform the labels into numeric values
Y = le.transform(Y)

This converts categorical class labels (e.g., 'Flu', 'Cold', 'Covid') into integers (e.g., 0, 1, 2), which are easier for most machine learning models to work with.

In [16]:
#dataset Splitting
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3, random_state=20)
# 30% for testing, 70% for training
#X_train: training features (70%)
#X_test: testing features (30%)

In [17]:
X_train.shape , X_test.shape , y_train.shape , y_test.shape

((3444, 132), (1476, 132), (3444,), (1476,))

### Models Selection and Training

In [18]:
#The reason for importing many models in this script is usually to perform a model comparison.
#that is, to train multiple classifiers on the same dataset and compare their performance.

# Create a dictionary to store models
models = {
    'SVC': SVC(kernel='linear'), #kernel='linear' means it assumes the data is linearly separable 
    'RandomForest': RandomForestClassifier(n_estimators=100, random_state=42), #n_estimators=100 means it builds 100 trees.
    'GradientBoosting': GradientBoostingClassifier(n_estimators=100, random_state=42),
    'KNeighbors': KNeighborsClassifier(n_neighbors=5),
    'MultinomialNB': MultinomialNB()
}

| Model     | Description |
|:----------|---------:|
| RF        | Ensemble of decision trees, robust and widely used.     |
| SVM       | 	Good for small to medium datasets with a clear margin of separation.   |
| GradientBoostingClassifier | 	Sequential boosting model, often more accurate. |
| KNeighborsClassifier | Instance-based, makes predictions based on closest neighbors. |
| MultinomialNB | Naive Bayes, good for text and count-based features. |


In [29]:
# Loop through the models, train, test, and print results
for model_name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)

    # Test the model
    predictions = model.predict(X_test)

    # Calculate accuracy
    accuracy = accuracy_score(y_test, predictions)
    print(f"{model_name} Accuracy: {accuracy:.4f}")

    # Calculate confusion matrix
    cm = confusion_matrix(y_test, predictions)
    print(f"{model_name} Confusion Matrix:")
    print(np.array2string(cm, separator=', ')) #is used to neatly print the confusion matrix (cm) as a formatted string.
    print("\n" + "="*40 + "\n")

SVC Accuracy: 1.0000
SVC Confusion Matrix:
[[40,  0,  0, ...,  0,  0,  0],
 [ 0, 43,  0, ...,  0,  0,  0],
 [ 0,  0, 28, ...,  0,  0,  0],
 ...,
 [ 0,  0,  0, ..., 34,  0,  0],
 [ 0,  0,  0, ...,  0, 41,  0],
 [ 0,  0,  0, ...,  0,  0, 31]]


RandomForest Accuracy: 1.0000
RandomForest Confusion Matrix:
[[40,  0,  0, ...,  0,  0,  0],
 [ 0, 43,  0, ...,  0,  0,  0],
 [ 0,  0, 28, ...,  0,  0,  0],
 ...,
 [ 0,  0,  0, ..., 34,  0,  0],
 [ 0,  0,  0, ...,  0, 41,  0],
 [ 0,  0,  0, ...,  0,  0, 31]]


GradientBoosting Accuracy: 1.0000
GradientBoosting Confusion Matrix:
[[40,  0,  0, ...,  0,  0,  0],
 [ 0, 43,  0, ...,  0,  0,  0],
 [ 0,  0, 28, ...,  0,  0,  0],
 ...,
 [ 0,  0,  0, ..., 34,  0,  0],
 [ 0,  0,  0, ...,  0, 41,  0],
 [ 0,  0,  0, ...,  0,  0, 31]]


KNeighbors Accuracy: 1.0000
KNeighbors Confusion Matrix:
[[40,  0,  0, ...,  0,  0,  0],
 [ 0, 43,  0, ...,  0,  0,  0],
 [ 0,  0, 28, ...,  0,  0,  0],
 ...,
 [ 0,  0,  0, ..., 34,  0,  0],
 [ 0,  0,  0, ...,  0, 41,  0],
 [ 0

For a perfect classifier:<br>
All non-zero values are on the diagonal.<br>
The diagonal values (like 40, 43, 28...) are the correct predictions for each class.<br>
All off-diagonal values are 0 (i.e., no misclassifications).

##### Training using SVM Model
we have selected SVM because it the most sutiable for our dataset.

In [ ]:
# selecting svc
svc = SVC(kernel='linear')
svc.fit(X_train,y_train)
ypred = svc.predict(X_test)
accuracy_score(y_test,ypred)

##### Saving the model after training

In [20]:
pickle.dump(svc,open('svc.pkl','wb'))
#This saves (dumps) the trained model object svc into a file called svc.pkl.
#svc: the trained model object you want to save.
#'svc.pkl': the filename where the model will be stored.
#'wb': write binary mode — required for storing objects in binary form.
# When you save model,You don't have to retrain the model every time you use it.

##### Loading model after saving

In [21]:
# load model
svc = pickle.load(open('svc.pkl','rb'))
#You don't have to retrain the model every time you use it.
#'svc.pkl' is the name of the file where your model was saved earlier using pickle.dump.
#'rb' stands for "read binary" mode — because the model was saved in binary form.

### Model Testing

In [22]:
# test 1:
print("predicted disease :",svc.predict(X_test.iloc[0].values.reshape(1,-1))) #reshaped array(1,132)
# X_test.iloc[0] – Selects the first row (the first test instance).
# values.reshape(1, -1) – Converts the row into a 2D array because scikit-learn expects input shape as (n_samples, n_features).
#Even for one sample, it must be reshaped.

print("Actual Disease :", y_test[0])

predicted disease : [40]
Actual Disease : 40


C:\Users\Dubai\anaconda\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


In [23]:
# test 2:
print("predicted disease :",svc.predict(X_test.iloc[100].values.reshape(1,-1)))
print("Actual Disease :", y_test[100])

predicted disease : [39]
Actual Disease : 39


C:\Users\Dubai\anaconda\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


### Loading Database and Linking it with Predicitions

In [27]:
sym_des = pd.read_csv("Database/symtoms_df.csv")
precautions = pd.read_csv("Database/precautions_df.csv")
workout = pd.read_csv("Database/workout_df.csv")
description = pd.read_csv("Database/description.csv")
medications = pd.read_csv('Database/medications.csv')
diets = pd.read_csv("Database/diets.csv")

##### helper funciton
This function gathers related information for a given disease from database tables.<br>
This function takes a disease name dis and returns relevant descriptive information.

| Variable | Definition |
|--|--|
|desc| gets the disease description from description table.|
|pre| gets a list of four precaution steps for the disease.|
|med| gets a list of recommended medications.|
|die| gets diet recommendations.|
|wrkout| gets suggested workouts or physical activities.|

In [33]:
def helper(disease):
    desc = description[description['Disease'] == disease]['Description']
    # go to the description table check if the value in Disease column = the predicted disease then return the Description.
    desc = " ".join([w for w in desc]) #convert a pandas Series (desc) into a single string. 
    #his list comprehension iterates over the Series and collects each string value into a list.
    
    pre = precautions[precautions['Disease'] == disease][['Precaution_1', 'Precaution_2', 'Precaution_3', 'Precaution_4']]
    pre = [col for col in pre.values] #produces a list with a single element — the row itself — like this
                                      #  ['Drink more water', 'Take rest', 'Avoid junk food', 'Consult doctor']

    med = medications[medications['Disease'] == disease]['Medication']
    med = [med for med in med.values]

    die = diets[diets['Disease'] == disease]['Diet']
    die = [die for die in die.values]

    wrkout = workout[workout['disease'] == disease] ['workout']


    return desc,pre,med,die,wrkout



##### The main function for prediction
 predict a disease given a set of symptoms using a trained ML model 

In [37]:
symptoms_dict = {'itching': 0, 'skin_rash': 1, 'nodal_skin_eruptions': 2, 'continuous_sneezing': 3, 'shivering': 4, 'chills': 5, 'joint_pain': 6, 'stomach_pain': 7, 'acidity': 8, 'ulcers_on_tongue': 9, 'muscle_wasting': 10, 'vomiting': 11, 'burning_micturition': 12, 'spotting_ urination': 13, 'fatigue': 14, 'weight_gain': 15, 'anxiety': 16, 'cold_hands_and_feets': 17, 'mood_swings': 18, 'weight_loss': 19, 'restlessness': 20, 'lethargy': 21, 'patches_in_throat': 22, 'irregular_sugar_level': 23, 'cough': 24, 'high_fever': 25, 'sunken_eyes': 26, 'breathlessness': 27, 'sweating': 28, 'dehydration': 29, 'indigestion': 30, 'headache': 31, 'yellowish_skin': 32, 'dark_urine': 33, 'nausea': 34, 'loss_of_appetite': 35, 'pain_behind_the_eyes': 36, 'back_pain': 37, 'constipation': 38, 'abdominal_pain': 39, 'diarrhoea': 40, 'mild_fever': 41, 'yellow_urine': 42, 'yellowing_of_eyes': 43, 'acute_liver_failure': 44, 'fluid_overload': 45, 'swelling_of_stomach': 46, 'swelled_lymph_nodes': 47, 'malaise': 48, 'blurred_and_distorted_vision': 49, 'phlegm': 50, 'throat_irritation': 51, 'redness_of_eyes': 52, 'sinus_pressure': 53, 'runny_nose': 54, 'congestion': 55, 'chest_pain': 56, 'weakness_in_limbs': 57, 'fast_heart_rate': 58, 'pain_during_bowel_movements': 59, 'pain_in_anal_region': 60, 'bloody_stool': 61, 'irritation_in_anus': 62, 'neck_pain': 63, 'dizziness': 64, 'cramps': 65, 'bruising': 66, 'obesity': 67, 'swollen_legs': 68, 'swollen_blood_vessels': 69, 'puffy_face_and_eyes': 70, 'enlarged_thyroid': 71, 'brittle_nails': 72, 'swollen_extremeties': 73, 'excessive_hunger': 74, 'extra_marital_contacts': 75, 'drying_and_tingling_lips': 76, 'slurred_speech': 77, 'knee_pain': 78, 'hip_joint_pain': 79, 'muscle_weakness': 80, 'stiff_neck': 81, 'swelling_joints': 82, 'movement_stiffness': 83, 'spinning_movements': 84, 'loss_of_balance': 85, 'unsteadiness': 86, 'weakness_of_one_body_side': 87, 'loss_of_smell': 88, 'bladder_discomfort': 89, 'foul_smell_of urine': 90, 'continuous_feel_of_urine': 91, 'passage_of_gases': 92, 'internal_itching': 93, 'toxic_look_(typhos)': 94, 'depression': 95, 'irritability': 96, 'muscle_pain': 97, 'altered_sensorium': 98, 'red_spots_over_body': 99, 'belly_pain': 100, 'abnormal_menstruation': 101, 'dischromic _patches': 102, 'watering_from_eyes': 103, 'increased_appetite': 104, 'polyuria': 105, 'family_history': 106, 'mucoid_sputum': 107, 'rusty_sputum': 108, 'lack_of_concentration': 109, 'visual_disturbances': 110, 'receiving_blood_transfusion': 111, 'receiving_unsterile_injections': 112, 'coma': 113, 'stomach_bleeding': 114, 'distention_of_abdomen': 115, 'history_of_alcohol_consumption': 116, 'fluid_overload.1': 117, 'blood_in_sputum': 118, 'prominent_veins_on_calf': 119, 'palpitations': 120, 'painful_walking': 121, 'pus_filled_pimples': 122, 'blackheads': 123, 'scurring': 124, 'skin_peeling': 125, 'silver_like_dusting': 126, 'small_dents_in_nails': 127, 'inflammatory_nails': 128, 'blister': 129, 'red_sore_around_nose': 130, 'yellow_crust_ooze': 131}
# index mapping: This dictionary assigns a unique numeric index to each symptom.
# This mapping ensures that each symptom corresponds to a fixed position (column) in a feature vector.

diseases_list = {15: 'Fungal infection', 4: 'Allergy', 16: 'GERD', 9: 'Chronic cholestasis', 14: 'Drug Reaction', 33: 'Peptic ulcer diseae', 1: 'AIDS', 12: 'Diabetes ', 17: 'Gastroenteritis', 6: 'Bronchial Asthma', 23: 'Hypertension ', 30: 'Migraine', 7: 'Cervical spondylosis', 32: 'Paralysis (brain hemorrhage)', 28: 'Jaundice', 29: 'Malaria', 8: 'Chicken pox', 11: 'Dengue', 37: 'Typhoid', 40: 'hepatitis A', 19: 'Hepatitis B', 20: 'Hepatitis C', 21: 'Hepatitis D', 22: 'Hepatitis E', 3: 'Alcoholic hepatitis', 36: 'Tuberculosis', 10: 'Common Cold', 34: 'Pneumonia', 13: 'Dimorphic hemmorhoids(piles)', 18: 'Heart attack', 39: 'Varicose veins', 26: 'Hypothyroidism', 24: 'Hyperthyroidism', 25: 'Hypoglycemia', 31: 'Osteoarthristis', 5: 'Arthritis', 0: '(vertigo) Paroymsal  Positional Vertigo', 2: 'Acne', 38: 'Urinary tract infection', 35: 'Psoriasis', 27: 'Impetigo'}
# dictionary maps numeric prediction output from the model to the actual disease label.

# Model Prediction function
def get_predicted_disease(patient_symptoms): #accept a list of patient symptoms.
    input_vector = np.zeros(len(symptoms_dict))  #vector with 132 values (zeros).
    for item in patient_symptoms: #iterate over each symptom in symptoms list.
        input_vector[symptoms_dict[item]] = 1 #Set ones at the positions of observed symptoms.
    return diseases_list[svc.predict([input_vector])[0]] #This passes the input vector to the trained SVC model.
    #Map predicted integer to disease label:diseases_list[4] → 'Allergy' and return Allergy
    #[0] means "take the first (and only) element of the result" → 4

In [ ]:
# do not run this cell
#more deeply explainaiton
'''
for item in patient_symptoms:
input_vector[symptoms_dict[item]] = 1
Suppose patient_symptoms = ['itching', 'skin_rash', 'vomiting']
symptoms_dict[item] => symptoms_dict['vomiting'] this equal 11.
so , input_vector[11] = 1 ← vomiting
and so on,
input_vector[0] = 1 ← itching
input_vector[1] = 1 ← skin_rash
The rest remain 0.

So the final vector looks like:

[1. 1. 0. 0. 0. ... 1. 0. ... 0.]
'''

In [35]:
# Test
# Split the user's input into a list of symptoms (assuming they are comma-separated) # itching,skin_rash,nodal_skin_eruptions
symptoms = input("Enter your symptoms.......")
user_symptoms = [s.strip() for s in symptoms.split(',')]
# Remove any extra characters, if any
user_symptoms = [symptom.strip("[]' ") for symptom in user_symptoms]
predicted_disease = get_predicted_disease(user_symptoms)

desc, pre, med, die, wrkout = helper(predicted_disease)

print("=================predicted disease============")
print(predicted_disease)
print("=================description==================")
print(desc)
print("=================precautions==================")
i = 1
for p_i in pre[0]:
    print(i, ": ", p_i)
    i += 1

print("=================medications==================")
for m_i in med:
    print(i, ": ", m_i)
    i += 1

print("=================workout==================")
for w_i in wrkout:
    print(i, ": ", w_i)
    i += 1

print("=================diets==================")
for d_i in die:
    print(i, ": ", d_i)
    i += 1


Enter your symptoms.......  itching,skin_rash,nodal_skin_eruptions


=================predicted disease============
Fungal infection
=================description==================
Fungal infection is a common skin condition caused by fungi.
=================precautions==================
1 :  bath twice
2 :  use detol or neem in bathing water
3 :  keep infected area dry
4 :  use clean cloths
=================medications==================
5 :  ['Antifungal Cream', 'Fluconazole', 'Terbinafine', 'Clotrimazole', 'Ketoconazole']
=================workout==================
6 :  Avoid sugary foods
7 :  Consume probiotics
8 :  Increase intake of garlic
9 :  Include yogurt in diet
10 :  Limit processed foods
11 :  Stay hydrated
12 :  Consume green tea
13 :  Eat foods rich in zinc
14 :  Include turmeric in diet
15 :  Eat fruits and vegetables
=================diets==================
16 :  ['Antifungal Diet', 'Probiotics', 'Garlic', 'Coconut oil', 'Turmeric']


C:\Users\Dubai\anaconda\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


In [38]:
# let's use pycharm flask app
# but install this version in pycharm because we have trained our model on this version.
import sklearn
print(sklearn.__version__)


1.5.1
